deps/rabbit/src/rabbit_binding.erl

```diff
@@ -100,204 +89,106 @@ recover(XNames, QNames) ->
                 end,
     {ok, Gatherer} = gatherer:start_link(),
-    [recover_semi_durable_route(Gatherer, R, SelectSet(Dst)) ||
-        R = #route{binding = #binding{destination = Dst}} <-
-            rabbit_misc:dirty_read_all(rabbit_semi_durable_route)],
+    rabbit_db_binding:recover(
+      fun(Binding, Src, Dst, Fun) ->
+              recover_semi_durable_route(Gatherer, Binding, Src, Dst, SelectSet(Dst), Fun)
+      end),
     empty = gatherer:out(Gatherer),
     ok = gatherer:stop(Gatherer),
     ok.
 
-recover_semi_durable_route(Gatherer, R = #route{binding = B}, ToRecover) ->
-    #binding{source = Src, destination = Dst} = B,
+recover_semi_durable_route(Gatherer, Binding, Src, Dst, ToRecover, Fun) ->
     case sets:is_element(Dst, ToRecover) of
         true  -> {ok, X} = rabbit_exchange:lookup(Src),
                  ok = gatherer:fork(Gatherer),
                  ok = worker_pool:submit_async(
                         fun () ->
-                                recover_semi_durable_route_txn(R, X),
+                                Fun(Binding, X),
                                 gatherer:finish(Gatherer)
                         end);
         false -> ok
     end.
 
-recover_semi_durable_route_txn(R = #route{binding = B}, X) ->
-    rabbit_misc:execute_mnesia_transaction(
-      fun () ->
-              case mnesia:read(rabbit_semi_durable_route, B, read) of
-                  [] -> no_recover;
-                  _  -> ok = sync_transient_route(R, should_index_table(X), fun mnesia:write/3),
-                        rabbit_exchange:serial(X)
-              end
-      end,
-      fun (no_recover, _)     -> ok;
-          (_Serial,    true)  -> x_callback(transaction, X, add_binding, B);
-          (Serial,     false) -> x_callback(Serial,      X, add_binding, B)
-      end).
-
 -spec exists(rabbit_types:binding()) -> boolean() | bind_errors().
 
 exists(#binding{source = ?DEFAULT_EXCHANGE(_),
                 destination = #resource{kind = queue, name = QName} = Queue,
                 key = QName,
                 args = []}) ->
     rabbit_amqqueue:exists(Queue);
-exists(Binding) ->
-    binding_action(
-      Binding, fun (_Src, _Dst, B) ->
-                       rabbit_misc:const(mnesia:read({rabbit_route, B}) /= [])
-               end, fun not_found_or_absent_errs/1).
+exists(Binding0) ->
+    Binding = sort_args(Binding0),
+    rabbit_db_binding:exists(Binding).
 
 -spec add(rabbit_types:binding(), rabbit_types:username()) -> bind_res().
 
 add(Binding, ActingUser) -> add(Binding, fun (_Src, _Dst) -> ok end, ActingUser).
 
 -spec add(rabbit_types:binding(), inner_fun(), rabbit_types:username()) -> bind_res().
 
-add(Binding, InnerFun, ActingUser) ->
-    binding_action(
-      Binding,
-      fun (Src, Dst, B) ->
-              case rabbit_exchange:validate_binding(Src, B) of
-                  ok ->
-                      lock_resource(Src, read),
-                      lock_resource(Dst, read),
```

Comment:
Those two locks seem to be missing in the converted code. Is this on purpose?

Reply:
Accidental loss